ndarray/src/main/java/org/tensorflow/ndarray/impl/sparse/SparseNdArray.java

```diff
@@ -0,0 +1,397 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.ndarray.impl.sparse;
+
+import org.tensorflow.ndarray.LongNdArray;
+import org.tensorflow.ndarray.NdArray;
+import org.tensorflow.ndarray.NdArrays;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.ndarray.StdArrays;
+import org.tensorflow.ndarray.buffer.DataBuffer;
+import org.tensorflow.ndarray.buffer.DataBuffers;
+import org.tensorflow.ndarray.impl.dimension.DimensionalSpace;
+import org.tensorflow.ndarray.impl.sparse.slice.ObjectSparseSlice;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * sparse array for the any data type
+ *
+ * <p>A sparse array has two separate dense arrays: indices, values, and a shape that represents the
+ * dense shape.
+ *
+ * <p><em>NOTE:</em> all Sparse Arrays are readonly for the {@link #set(NdArray, long...)} and
+ * {@link #setObject(Object, long...)} methods
+ *
+ * <pre>{@code
+ * SparseNdArray<String> st = new SparseNdArray<>(
+ *      StdArrays.of(new long[][] {{0, 0}, {1, 2}}),
+ *      NdArrays.vectorOf("first", "second"),
+ *      Shape.of(3, 4));
+ *
+ * }</pre>
+ *
+ * <p>represents the dense array:
+ *
+ * <pre>{@code
+ * [[true, false, false, false]
+ *  [false, false, true, false]
+ *  [false, false, false, false]]
+ *
+ * }</pre>
+ */
+public class SparseNdArray<T, U extends NdArray<T>> extends AbstractSparseNdArray<T, U>
+    implements org.tensorflow.ndarray.SparseNdArray<T, U> {
```

Comment:
The name conflict between the interface and the "object" implementation is a bit annoying but I'm ok keeping it like that for now and readjust later if needed, since that `SparseNdArray` is a "private" implementation.